Update the `check_schema_delta` script to account for when the schema version has been bumped locally

H-Shay opened the pull request:
Previously, the script used the schema version on develop as the canonical version, so if your PR bumped the schema version and put schema delta files in the folder corresponding to the new schema version it would fail the CI check.

This PR addresses that by checking to see if the schema version is changed locally, and if it is, uses the local version as the canonical schema version number. 

I'm new to the python git library so there may be a more elegant way to do this but I did test it locally and afaict it does work.

DMRobertson commented on scripts-dev/check_schema_delta.py:
To summarise,  we previously enforced (and continue to enforce) the invariant

> PRs can only introduce migrations for the current schema

but we have now altered "current schema" to mean the PR's current schema version, falling back to develop's schema version.

I think we ought to have a check which enforces that the PR's schema version is either (develop's schema version) or (develop's schema version + 1). That would prevent the following footgun:

- Develop is on schema version 1.
- Someone opens PR X which bumps the schema version from 1 to 2. We ignore it for ages.
- Develop sees several schema updates. It's now on schema 5.
- We rerun the CI for PR X which still targets version 2. 
- It passes.
- :warning: We merge in new migrations to schema delta 2, after it should have been frozen!

DMRobertson reviewed (changes requested):
See comment. (Do you agree with my reasoning?)

H-Shay commented:
Makes sense to me and I even think you mentioned this in the meeting but it didn't fully register for some reason - I will update this to implement the check you've outlined, thanks for taking a look!

H-Shay commented:
I've added a check that determines if the proposed schema version is +/- 1 from the schema version on develop (-1 I added so it doesn't irritatingly break when something needs to be reverted) before then using the local version as the canonical version if it passes. Hopefully this addresses your concern?

erikjohnston commented on scripts-dev/check_schema_delta.py:
```diff
@@ -40,10 +40,39 @@ def main(force_colors: bool) -> None:
+            # we've changed the schema file, let's check if the version has been changed
+            res = repo.git.show(
+                f"{repo.active_branch}:synapse/storage/schema/__init__.py"
+            )
```
I think instead of looping through the diff we can just read the local file in the checkout? Which will be the merge commit?

DMRobertson reviewed (approved):
I suspect there are edge cases around all this (e.g. reverting the schema to version N-1 and adding new migrations to schema N - 1 is permitted by this check, but would be Very Bad if we ever released a Synapse with schema version N).

But this doesn't have to perfectly encapsulate all of the schema constraints to be useful. :+1:

(Have you tried merging this into your other PR to check that its CI would then run after this change?)

H-Shay commented:
Merged #15466 into develop.
